pkg/apis/replication/v1alpha/endpoint_spec.go

```diff
@@ -0,0 +1,111 @@
+
+	"github.com/arangodb/kube-arangodb/pkg/util"
+	"github.com/arangodb/kube-arangodb/pkg/util/k8sutil"
+	"github.com/pkg/errors"
+)
+
+// EndpointSpec contains the specification used to reach the syncmasters
+// in either source or destination mode.
+type EndpointSpec struct {
+	// DeploymentName holds the name of an ArangoDeployment resource.
+	// If set this provides default values for masterEndpoint, auth & tls.
+	DeploymentName *string `json:"deploymentName,omitempty"`
+	// MasterEndpoints holds a list of URLs used to reach the syncmaster(s).
+	MasterEndpoint []string `json:"masterEndpoint,omitempty"`
+	// Authentication holds settings needed to authentication at the syncmaster.
+	Authentication EndpointAuthenticationSpec `json:"auth"`
+	// TLS holds settings needed to verify the TLS connection to the syncmaster.
+	TLS EndpointTLSSpec `json:"tls"`
+}
+
+// GetDeploymentName returns the value of deploymentName.
+func (s EndpointSpec) GetDeploymentName() string {
+	return util.StringOrDefault(s.DeploymentName)
+}
+
+// HasDeploymentName returns the true when a non-empty deployment name it set.
+func (s EndpointSpec) HasDeploymentName() bool {
+	return s.GetDeploymentName() != ""
+}
+
+// Validate the given spec, returning an error on validation
+// problems or nil if all ok.
+func (s EndpointSpec) Validate() error {
+	if err := k8sutil.ValidateOptionalResourceName(s.GetDeploymentName()); err != nil {
+		return maskAny(err)
+	}
+	for _, ep := range s.MasterEndpoint {
+		if _, err := url.Parse(ep); err != nil {
+			return maskAny(errors.Wrapf(ValidationError, "Invalid master endpoint '%s': %s", ep, err))
+		}
+	}
+	hasDeploymentName := s.HasDeploymentName()
+	if !hasDeploymentName && len(s.MasterEndpoint) == 0 {
+		return maskAny(errors.Wrapf(ValidationError, "Provide a deploy name or at least one master endpoint"))
+	}
+	if err := s.Authentication.Validate(!hasDeploymentName); err != nil {
+		return maskAny(err)
+	}
+	if err := s.TLS.Validate(!hasDeploymentName); err != nil {
+		return maskAny(err)
+	}
+	return nil
+}
+
+// SetDefaults fills empty field with default values.
+func (s *EndpointSpec) SetDefaults() {
+	s.Authentication.SetDefaults()
+	s.TLS.SetDefaults()
+}
+
+// SetDefaultsFrom fills empty field with default values from the given source.
+func (s *EndpointSpec) SetDefaultsFrom(source EndpointSpec) {
+	if s.DeploymentName == nil {
+		s.DeploymentName = util.NewStringOrNil(source.DeploymentName)
+	}
+	s.Authentication.SetDefaultsFrom(source.Authentication)
+	s.TLS.SetDefaultsFrom(source.TLS)
+}
+
+// ResetImmutableFields replaces all immutable fields in the given target with values from the source spec.
+// It returns a list of fields that have been reset.
+// Field names are relative to `spec.`.
+func (s EndpointSpec) ResetImmutableFields(target *EndpointSpec, fieldPrefix string) []string {
+	var result []string
+	if s.GetDeploymentName() != target.GetDeploymentName() {
+		result = append(result, fieldPrefix+"deploymentName")
```

Comment:
should "deploymentName" be pulled off the struct via reflection?  This code appears to create two places where tag names are defined.  If one place changes, will coder see second place too?  My comment applies to "auth." and "xls." below.

Reply:
It could be, but that is (1) expensive and (2) really annoying code to look at.
I would prefer to make it a constant, but using constants in tags in not possible.
In the code here it is only for showing the user, so there is not critical semantics.